supertokens_python/recipe/emailverification/recipe.py

```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
             "Initialisation not done. Did you forget to call the SuperTokens.init function?"
         )
 
+    @staticmethod
+    def get_instance_optional() -> Optional[EmailVerificationRecipe]:
+        return EmailVerificationRecipe.__instance
+
     @staticmethod
     def reset():
         if ("SUPERTOKENS_ENV" not in environ) or (
             environ["SUPERTOKENS_ENV"] != "testing"
         ):
             raise_general_exception("calling testing function in non testing env")
         EmailVerificationRecipe.__instance = None
+
+    async def get_email_for_user_id(
+        self, user_id: str, user_context: Dict[str, Any]
+    ) -> Union[GetEmailForUserIdOkResult, EmailDoesnotExistError, UnknownUserIdError]:
+        if self.config.get_email_for_user_id is not None:
+            res = await self.config.get_email_for_user_id(user_id, user_context)
+            if not isinstance(res, UnknownUserIdError):
+                return res
+
+        for f in self.get_email_for_user_id_funcs_from_other_recipes:
+            res = await f(user_id, user_context)
+            if not isinstance(res, UnknownUserIdError):
+                return res
+
+        return UnknownUserIdError()
+
+    def add_get_email_for_user_id_func(self, f: Callable[[str, Dict[str, Any]], Any]):
+        self.get_email_for_user_id_funcs_from_other_recipes.append(f)
+
+
+class EmailVerificationClaimClass(BooleanClaim):
+    def __init__(self):
+        async def fetch_value(user_id: str, user_context: Dict[str, Any]) -> bool:
+            recipe = EmailVerificationRecipe.get_instance()
+            email_info = await recipe.get_email_for_user_id(user_id, user_context)
+
+            if isinstance(email_info, GetEmailForUserIdOkResult):
+                return await recipe.recipe_implementation.is_email_verified(
+                    user_id, email_info.email, user_context
+                )
+            if isinstance(email_info, EmailDoesnotExistError):
+                # we consider people without email addresses as validated
+                return True
+            raise Exception(
+                "Should never come here: UNKNOWN_USER_ID or invalid result from get_email_for_user"
+            )
+
+        super().__init__("st-ev", fetch_value)
+
+        self.validators = EmailVerificationClaimValidators(claim=self)
+
+
+EmailVerificationClaim = EmailVerificationClaimClass()
+
+
+class APIImplementation(APIInterface):
```

Comment:
why is this moved into this file and not kept in the other file?

Reply:
There was a cyclic import error that kept me stuck for hours. I still don't know how to fix it. Generally, this happens because of types so we just use `from _future__  import annotations` and `from typing import TYPE_CHECKING`.

But here we are using using the class (not just its type) so the error is really hard to get rid of. I have struggled with it for hours. The only feasible solution for now was to move it here. Need some help to fix it.